tezos/new_context/src/working_tree/storage.rs

```diff
@@ -0,0 +1,574 @@
+    #[skip]
+    __: B14,
+    start: B30,
+    length: B20,
+}
+
+impl Default for TreeStorageId {
+    fn default() -> Self {
+        Self::empty()
+    }
+}
+
+impl TreeStorageId {
+    fn new_tree(start: usize, end: usize) -> Result<Self, StorageIdError> {
+        let length = end
+            .checked_sub(start)
+            .ok_or(StorageIdError::TreeInvalidStartEnd)?;
+
+        if start & !0x3FFFFFFF != 0 {
+            // Must fit in 30 bits
+            return Err(StorageIdError::TreeStartTooBig);
+        }
+
+        if length & !0xFFFFF != 0 {
+            // Must fit in 20 bits
+            return Err(StorageIdError::TreeLengthTooBig);
+        }
+
+        let tree_id = Self::new()
+            .with_start(start as u32)
+            .with_length(length as u32);
+
+        debug_assert_eq!(tree_id.get(), (start as usize, end));
+
+        Ok(tree_id)
+    }
+
+    fn get(self) -> (usize, usize) {
+        let start = self.start() as usize;
+        let length = self.length() as usize;
+
+        (start, start + length)
+    }
+
+    pub fn empty() -> Self {
+        // Never fails
+        Self::new_tree(0, 0).unwrap()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.length() == 0
+    }
+}
+
+impl From<TreeStorageId> for u64 {
+    fn from(tree_id: TreeStorageId) -> Self {
+        let bytes = tree_id.into_bytes();
+        u64::from_ne_bytes(bytes)
+    }
+}
+
+impl From<u64> for TreeStorageId {
+    fn from(entry_id: u64) -> Self {
+        Self::from_bytes(entry_id.to_ne_bytes())
+    }
+}
+
+#[derive(Debug)]
+pub enum StorageIdError {
+    BlobSliceTooBig,
+    BlobStartTooBig,
+    BlobLengthTooBig,
+    TreeInvalidStartEnd,
+    TreeStartTooBig,
+    TreeLengthTooBig,
+    NodeIdError,
+    StringNotFound,
+    TreeNotFound,
+    BlobNotFound,
+    NodeNotFound,
+}
+
+impl From<NodeIdError> for StorageIdError {
+    fn from(_: NodeIdError) -> Self {
+        Self::NodeIdError
+    }
+}
+
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BlobStorageId {
+    /// Note: Must fit in NodeInner.entry_id (61 bits)
+    ///
+    /// | 3 bits  | 1 bit     | 60 bits |
+    /// |---------|-----------|---------|
+    /// | empty   | is_inline | value   |
+    ///
+    /// value inline:
+    /// | 4 bits | 56 bits |
+    /// |--------|---------|
+    /// | length | value   |
+    ///
+    /// value not inline:
+    /// | 32 bits | 28 bits |
+    /// |---------|---------|
+    /// | start   | length  |
+    bits: u64,
+}
+
+impl From<BlobStorageId> for u64 {
+    fn from(blob_id: BlobStorageId) -> Self {
+        blob_id.bits
+    }
+}
+
+impl From<u64> for BlobStorageId {
+    fn from(entry: u64) -> Self {
+        Self { bits: entry }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum BlobRef {
+    Inline { length: u8, value: [u8; 7] },
+    Ref { start: usize, end: usize },
+}
+
+impl BlobStorageId {
+    pub fn new_inline(value: &[u8]) -> Result<Self, StorageIdError> {
```

Comment:
@sebastiencs these two would be better named `try_new_inline` and `try_new` to make it explicit that they can fail.